src/main/golem/util/validation/common.kt

```diff
@@ -0,0 +1,172 @@
+    operator fun invoke() : ValidationContext = this
+}
+
+
+/**
+ * Return a validation context that can be used to validate the given matrix with the default name of "matrix".
+ *
+ * In general it is preferable to use the other verison of this function that lets you specify a name. All
+ * matrices that are validated with this method will appear in the error message with the name "matrix"
+ *
+ * @param matrix A matrix to validate.
+ * @returns A validation context that can be used to validate the given matrix.
+ */
+public fun testMatrix(matrix: Matrix<Double>) : ValidationContext = testMatrix(matrix, DEFAULT_NAME)
+
+/**
+ * Return a validation context that can be used to validate the given matrix with the given name.
+ *
+ * <code>
+ * testMatrix(myMatrix, "myMatrix").dim(1, 4).validate() // validates myMatrix is 1 by 4.
+ * </code>
+ *
+ * @param matrix A matrix to validate.
+ * @param name The name of the matrix (used in displayed errors)
+ * @returns A validation context that can be used to validate the given matrix.
+ */
+public fun testMatrix(matrix: Matrix<Double>, name: String) : ValidationContext  {
+    val ctx = ValidationContext()
+    ctx.matrixNames.add(name)
+    ctx.matrices.add(matrix)
+    return ctx
+}
+
+/**
+ * Execute the given rules within a ValidationContext, letting you validate multiple matrices at once with
+ * interrelated dimensions, and return a list of matrices that match your validation rules.
+ *
+ * <pre>
+ * {@code
+ * val (myFoo, myBar) = validate {
+ *     foo("foo") { 1 x 'N' }
+ *     bar("bar") { 'N' x 1; max = 0.5 }
+ * }
+ * }
+ * </pre>
+ */
+public fun validate(fn: ValidationContext.() -> Unit) {
+    val ctx = ValidationContext()
+    fn(ctx)
+    ctx.validate()
+}
+
+/**
+ * Use the given fn to validate a matrix. Return either the matrix itself or a copy that has been transformed
+ * to match the validation rules.
+ *
+ * In general it is preferable to use the other verison of this function that lets you specify a name. All
+ * matrices that are validated with this method will appear in the error message with the name "matrix"
+ *
+ * @see Matrix<Double>.validate(name, fn)
+ * @param fn Validation rules for the matrix.
+ * @returns Either a reference to the input matrix itself, or a transformed version. The return value is
+ * guaranteed to pass the validation rules.
+ */
+public fun Matrix<Double>.validate(fn: ValidationContext.() -> Unit) = validate(DEFAULT_NAME, fn)
+
+
+/**
+ * Use the given fn to validate a matrix with the given name. Return either the matrix itself or a copy that
+ * has been transformed to match the validation rules.
+ *
+ * <code>
+ * myMatrix.validate("myMatrix") { 1 x 4 } // validates myMatrix is 1 by 4.
+ * </code>
+ *
+ * @param name The name of the matrix (used in displayed errors)
+ * @param fn Validation rules for the matrix.
+ */
+public fun Matrix<Double>.validate(name: String, fn: ValidationContext.() -> Unit) {
```

Comment:
Theres an escape hatch "getDouble" which breaks encapsulation that could be used (other numerical typed matrices will be cast). When Java 10 hits we'll hopefully be able to do something more clever and typesafe here (i.e. subclass a value type with a common interface with add, greaterThan, etc.)